### PR TITLE
[ctx_prof] Fix `ProfileAnnotator::allTakenPathsExit`

### PR DESCRIPTION
**Pull request by mtrofin**
Added tests to the validator and fixed issues stemming from the previous skipping over BBs with single successors - which is incorrect. That would be now picked by added tests where the assertions are expected to be triggered.

### CODE REVIEW

**Comment by mtrofin** [comment]
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp [hunk 1]
```diff
@@ -233,28 +233,37 @@ class ProfileAnnotator final {
...
+        continue;
+      }
+      if (succ_size(BB) == 1) {
+        llvm::append_range(Worklist, successors(BB));
+        continue;
+      }
       const auto &BBInfo = getBBInfo(*BB);
-      bool Inserted = false;
+      bool HasAWayOut = false;
       for (auto I = 0U; I < BB->getTerminator()->getNumSuccessors(); ++I) {
```
Comment (teresajohnson) [inline llvm/lib/Transforms/Instrumentation/PGOCtxProfFlattening.cpp:254]: can this use succ_size as used above, for consistency? And maybe call that once and save the result.

File: llvm/test/Analysis/CtxProfAnalysis/flatten-check-path.ll [hunk 2]
```diff
@@ -0,0 +1,85 @@
+; REQUIRES: asserts
+; Check that the profile annotator works: we hit an exit and non-zero paths to
+; already visited blocks are not disconsidered.
```
Comment (teresajohnson) [inline llvm/test/Analysis/CtxProfAnalysis/flatten-check-path.ll:3]: Not sure what you are trying to say by "are not disconsidered" here, especially with the double-negative.

**Review by teresajohnson — APPROVED** [review]
A couple of small comments but otherwise lgtm

**Comment by mtrofin** [comment]
* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
* **#109203** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109185** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109184** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109184?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109183** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109183?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mtrofin and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by mtrofin** [comment]
### Merge activity

* **Sep 19, 12:06 AM EDT**: @mtrofin started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/109183).

**Pull request merged**